packages/cozy-realtime/src/index.js

```diff
@@ -1,339 +1,320 @@
+   * - create socket
+   * - listen cozyClient events
+   * - unsubscribeAll if window unload
+   *
+   * @constructor
+   * @param {CozyClient} cozyClient  A cozy client
+   */
+  constructor({ cozyClient }) {
+    this._cozyClient = cozyClient
+
+    this._updateAuthentication = this._updateAuthentication.bind(this)
+    this.unsubscribeAll = this.unsubscribeAll.bind(this)
+    this._receiveMessage = this._receiveMessage.bind(this)
+    this._receiveError = this._receiveError.bind(this)
+    this._resubscribe = this._resubscribe.bind(this)
+    this._beforeUnload = this._beforeUnload.bind(this)
+    this._resetSocket = this._resetSocket.bind(this)
+
+    this._createSocket()
+
+    this._cozyClient.on('login', this._updateAuthentication)
+    this._cozyClient.on('tokenRefreshed', this._updateAuthentication)
+    this._cozyClient.on('logout', this.unsubscribeAll)
+
+    if (global) {
+      global.addEventListener('beforeunload', this._beforeUnload)
+      global.addEventListener('online', this._resubscribe)
+      global.removeEventListener('offline', this._resetSocket)
+    }
   }
 
-  const protocol = options.secure ? 'wss:' : 'ws:'
-  const domain = options.domain || new URL(options.url).host
+  _beforeUnload() {
+    global.removeEventListener('beforeunload', this._windowUnload)
+    this.unsubscribeAll()
+  }
 
-  if (!domain) {
-    throw new Error('Unable to detect domain')
+  /**
+   * Create a Socket with cozyClient credential
+   */
+  _createSocket() {
+    if (!this._socket) {
+      const getUrl = () => getWebSocketUrl(this._cozyClient)
+      const getToken = () => getWebSocketToken(this._cozyClient)
+
+      this._socket = new Socket(getUrl, getToken)
+      this._socket.on('message', this._receiveMessage)
+      this._socket.on('error', this._receiveError)
+    }
   }
 
-  const socket = new WebSocket(
-    `${protocol}//${domain}/realtime/`,
-    'io.cozy.websocket'
-  )
+  /**
+   * When socket send error it test to reconnect
+   */
+  _receiveError(error) {
+    logger.info(`Receive error: ${error}`)
 
-  const windowUnloadHandler = () => socket.close()
-  window.addEventListener('beforeunload', windowUnloadHandler)
+    this._resetSocket()
 
-  socket.onmessage = onmessage
-  socket.onclose = event => {
-    window.removeEventListener('beforeunload', windowUnloadHandler)
-    if (typeof onclose === 'function') onclose(event, numRetries, retryDelay)
-  }
-  socket.onerror = error => console.error(`WebSocket error: ${error.message}`)
-
-  socketPromise = new Promise(resolve => {
-    socket.onopen = () => {
-      socket.send(
-        JSON.stringify({
-          method: 'AUTH',
-          payload: options.token
-        })
-      )
-      resolve(socket)
+    if (this._retryLimit === 0) {
+      this.emit('error', error)
+    } else {
+      if (this.retry) {
+        clearTimeout(this.retry)
+      }
+      if (global.navigator.onLine) {
+        this.retry = setTimeout(this._resubscribe, this._retryDelay)
```

Comment:
Maybe we should keep trying here even if we are not online. Who knows maybe the connection will get back soon.

Reply:
[Here we were waiting twice the previous delay](https://github.com/cozy/cozy-libs/pull/355/files#diff-1cff02e65335a8e39ac841204ef1b280R242). Maybe twice was a lot but what made you remove this code ?

Reply:
On devrait aussi ré-écouter si on est de nouveau `online` pour `retry` ? `navigator.onLine` n'est pas très fiable (enfin si mais on peut être en ligne sans avoir accès à internet). On pourrait aussi tester si on est en mode cordova et faire appel à network info pour écouter les events de connexion et lancer le boulot côté realtime

Reply:
C'est le cas actuellement, on relance les tout lorsque l'on revient en online voir le code https://github.com/cozy/cozy-libs/pull/424/files#diff-29f088a019f40d2242078e1052efc960R147